Devise

IgorPgrm opened the pull request.

IgorPgrm committed:
+gem Devise install

IgorPgrm committed:
Настройка Devise

IgorPgrm committed:
Формы входа и регистрации, гем для подтверждения емайла

IgorPgrm committed:
+ActionMailer, шаблоны, вызов отправки письма после прохождения теста

IgorPgrm committed:
Редактирование и создание тестов перенесено в Admin::

IgorPgrm committed:
+Админ панель, +стили
Редактирование доступно только для администраторов

IgorPgrm committed:
+Перенаправление админа на админ-панель

IgorPgrm committed:
+Назначение автором теста, администратора который его создал

IgorPgrm committed:
Автор поста админ, переписан метод

IgorPgrm committed:
+flash Приветствие пользователя после входа

IgorPgrm committed:
Исправлено приветствие после входа
Форма регистрации исправлен тип полей

BubuntuClu commented on app/controllers/admin/base_controller.rb:
```diff
@@ -0,0 +1,11 @@
+  private
+
+  def admin_required!
+    redirect_to root_path, alert: 'You are not authorized to view this page' unless current_user.is_a?(Admin)
```
добавь метод в модель `User` который будет проверять является ли пользователь админом. `admin?`

BubuntuClu commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    @test = Test.new
+  end
+
+  def start
```
старт в админке не нужен

BubuntuClu commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    end
+  end
+
+  def search
```
что ищем?

BubuntuClu commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -0,0 +1,54 @@
+
+  private
+  def test_params
+    params.require(:test).permit(:title, :level, :category_id, :author_id)
```
author_id лишний параметр

BubuntuClu commented on app/controllers/application_controller.rb:
```diff
@@ -3,26 +3,18 @@
-      cookies[:redirect_path] = request.path
-      redirect_to login_path, alert: 'Are you a Guru? Verify your e-mail and password please'
+  def after_sign_in_path_for(resource)
+    if resource.is_a?(Admin)
```
`resource.admin?`

BubuntuClu replied:
можно выводить сообщение для всех и тогда будет что-то типа такого
```
def
  flash[:notice]= ...
  resource.admin? ? admin_tests_path : tests_path
end
```

BubuntuClu commented on app/controllers/application_controller.rb:
```diff
@@ -3,26 +3,18 @@
-
-  def logged_in?
-    current_user.present?
+  def rescue_record_not_found
```
мы же с тобой решили что это плохой метод и он путает. его надо убрать

BubuntuClu commented on app/controllers/tests_controller.rb:
```diff
@@ -1,7 +1,7 @@
 class TestsController < ApplicationController
   before_action :authenticate_user!
-  before_action :find_test, only: %i[show edit update destroy start]
+  before_action :find_test, only: %i[start]
```
ради одного экшена не стоит колбэк делать. вынеси тогда код в сам экшен

BubuntuClu commented on app/controllers/tests_controller.rb:
```diff
@@ -1,7 +1,7 @@
+  before_action :find_test, only: %i[start]
 
   def index
     @tests = Test.all
```
экшен new тут тоже не нужен

BubuntuClu commented on db/migrate/20201224203735_add_devise_to_users.rb:
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+class AddDeviseToUsers < ActiveRecord::Migration[6.0]
```
эта миграция работает с таблицей `users`. так же миграции AddTypeAttributeToUsers и AddDefaultToUsers тоже с ней работают. во всех трех у тебя блоки `up/down`. объедини их в одну миграцию, зачем тебе три для изменения одной таблицы?

IgorPgrm committed:
Исправления
Проверка на админа
Удалены ненужные методы
Объединены миграции
Вывод приветствия для всех вошедших на сайт
